### PR TITLE
Implemented Factory Reset server for mock target

### PR DESCRIPTION
**Pull request by ericm**
Related issue: #98 
The strategy is to allow the target to pass in the Settings object which will tell the server what cases it should error on. These could be passed in as flags or otherwise when it's added to the CLI.

Also added proto definition and pb.go's

### CODE REVIEW

File: gnoi/gnoi.go [hunk 1]
```diff
@@ -35,9 +36,12 @@ var (
 
 // Server does blah.
```
Comment (loriscn12) [inline gnoi/gnoi.go:37]: remove

File: gnoi/gnoi.go [hunk 2]
```diff
@@ -35,9 +36,12 @@ var (
...
 type Server struct {
+	// Certs
```
Comment (loriscn12) [inline gnoi/gnoi.go:39]: avoid unnecessary comments

**Comment by ericm** [comment]
Added a unit test. Sorry if it makes the diff a bit long. Will keep them shorter in future

File: gnoi/gnoi.go [hunk 3]
```diff
@@ -35,9 +36,12 @@ var (
...
 	certServer         *cert.Server
 	certManager        *cert.Manager
 	defaultCertificate *tls.Certificate
+	// Factory Reset
```
Comment (loriscn12) [inline gnoi/gnoi.go:43]: same here, avoid unnecessary comments

File: gnoi/gnoi.go [hunk 4]
```diff
@@ -35,9 +36,12 @@ var (
...
+	resetServer *reset.Server
 }
 
 // NewServer does blah.
```
Comment (loriscn12) [inline gnoi/gnoi.go:47]: remove

File: gnoi/gnoi.go [hunk 5]
```diff
@@ -33,14 +34,15 @@ var (
 	rsaBitSize = 2048
 )
 
-// Server does blah.
+// Server represents a target
```
Comment (samribeiro) [inline gnoi/gnoi.go:37]: Thanks for cleaning this up. Go comments should be full sentences ending with a period.

File: gnoi/gnoi.go [hunk 6]
```diff
@@ -33,14 +34,15 @@ var (
...
 type Server struct {
 	certServer         *cert.Server
 	certManager        *cert.Manager
 	defaultCertificate *tls.Certificate
+	resetServer        *reset.Server
 }
 
-// NewServer does blah.
+// NewServer returns a new server that can be used by the mock target
```
Comment (samribeiro) [inline gnoi/gnoi.go:45]: use a period.

File: gnoi/gnoi.go [hunk 7]
```diff
@@ -93,6 +98,17 @@ func (s *Server) PrepareAuthenticated() *grpc.Server {
 	return grpc.NewServer(opts...)
 }
 
+// Register all implemented gRPC services
+func (s *Server) Register(g *grpc.Server) {
+	s.RegFactoryReset(g)
+	s.RegCertificateManagement(g)
+}
+
+// RegFactoryReset registers the gRPC server
```
Comment (samribeiro) [inline gnoi/gnoi.go:107]: period

File: gnoi/reset/server.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	osUnsupported bool
+}
+
+// Server for factory_reset service
+type Server struct {
+	pb.FactoryResetServer
+	*Settings
+}
+
+// NewServer generates a new factory reset server
```
Comment (samribeiro) [inline gnoi/reset/server.go:22]: period, add to all other comments.

File: gnoi/reset/server.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	"context"
+
+	"github.com/google/gnxi/gnoi/reset/pb"
+	"google.golang.org/grpc"
+)
+
+// Settings for configurable options in Server
+type Settings struct {
+	errorIfZero   bool
```
Comment (samribeiro) [inline gnoi/reset/server.go:12]: ZeroFillUnsupported bool
FactoryOSUnsupported bool

File: gnoi/reset/server.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+		if resError.ResetError != nil {
+			resError.ResetError.FactoryOsUnsupported = true
+		} else {
+			resError.ResetError = &pb.ResetError{FactoryOsUnsupported: true}
+		}
+	}
+	if resError.ResetError != nil {
+		return &pb.StartResponse{Response: resError}, nil
+	}
+	return &pb.StartResponse{Response: &pb.StartResponse_ResetSuccess{}}, nil
```
Comment (samribeiro) [inline gnoi/reset/server.go:48]: '''
resetError := &pb.ResetError{}
resetError.ZeroFillUnsupported = req.ZeroFill && s.ZeroFillUnsupported
resetError.FactoryOsUnsupported = req.FactoryOs && s.FactoryOSUnsupported

if resetError.ZeroFillUnsupported || resetError.FactoryOsUnsupported {
    return &pb.StartResponse{Response: &pb.StartResponse_ResetError{ResetError: resetError}}, nil
}

// TODO: Actually reset the system, example: certificates.

return &pb.StartResponse{Response: &pb.StartResponse_ResetSuccess{}}, nil
'''

File: gnoi/reset/server_test.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+						test.errorIfZero,
+						test.osUnsupported,
+						testReq.ZeroFill,
+						testReq.FactoryOs,
+						response)
+				}
+			}
+		}
+	}
+}
```
Comment (loriscn12) [inline gnoi/reset/server_test.go:54]: I think this works, however the double iteration makes it hard to read. I was thinking you could use t.Run and do something like the following (context: https://gobyexample.com/testing):


```suggestion
var tests []struct{
	name string
	request *pb.StartRequest
	settings Settings
}

func makeTests() tests{
	reqs := []pb.StartRequest{
		{ZeroFill: true},
		{ZeroFill: true, FactoryOs: true},
		{FactoryOs: true},
		{},
	}
	sets := []Settings{
		{false, false},
		{false, true},
		{true, false},
		{true, true},
	}
	t := tests
	for _, set := range sets {
		for _, req := range reqs {
		        name = <compute a name based on the current values>
			t = append(t, {name: "some name", request: req, settings: set})
		}
	}
	return t
}

tests := makeTests()
for _, test := range tests {
	s := NewServer(&test)
	t.Run(test.name, func(t *testing.T) {
		resp, err := s.Start(context.Background(), &testReq)
		if err == nil {
			t.Fatalf("ops")
		}
		switch response := resp.Response.(type) {
		case *pb.StartResponse_ResetSuccess:
			if test.errorIfZero && testReq.ZeroFill || test.osUnsupported && testReq.FactoryOs {
				t.Errorf(
					"Error occured on case: \nSettings{errIfZero:%v,osUnsupported:%v}\nRequest{ZeroFill:%v,FactoryOs:%v} \nResponseSuccess{%v},",
					test.errorIfZero,
					test.osUnsupported,
					testReq.ZeroFill,
					testReq.FactoryOs,
					response)
			}
		case *pb.StartResponse_ResetError:
			if response.ResetError.ZeroFillUnsupported != (test.errorIfZero && testReq.ZeroFill) ||
				response.ResetError.FactoryOsUnsupported != (test.osUnsupported && testReq.FactoryOs) {
				t.Errorf(
					"Error occured on case: \nSettings{errIfZero:%v,osUnsupported:%v}\nRequest{ZeroFill:%v,FactoryOs:%v} \nResponseError{%v},",
					test.errorIfZero,
					test.osUnsupported,
					testReq.ZeroFill,
					testReq.FactoryOs,
					response)
			}
		}
	})
}
```

Disclaimer, I haven't tested this :)

optional:
On top of my suggestion I would encourage you to have a want field inside your test struct and compute its value inside the make Tests func. Rationale: We should try to keep the actual test  as clean as possible.

File: gnoi/reset/server.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+func (s *Server) Start(ctx context.Context, req *pb.StartRequest) (*pb.StartResponse, error) {
+	resetError := &pb.ResetError{}
+	resetError.ZeroFillUnsupported = req.ZeroFill && s.ZeroFillUnsupported
+	resetError.FactoryOsUnsupported = req.FactoryOs && s.FactoryOSUnsupported
+
+	if resetError.ZeroFillUnsupported || resetError.FactoryOsUnsupported {
+		return &pb.StartResponse{Response: &pb.StartResponse_ResetError{ResetError: resetError}}, nil
+	}
+
+	// TODO: Reset the target device
```
Comment (samribeiro) [inline gnoi/reset/server.go:57]: my fault =)

// TODO: Reset the Target.

**Comment by ericm** [comment]
No problem. Just changed it :)

**Comment by samribeiro** [comment]
seems there are some diffs in the protos used? Would you mind resolving them?

**Comment by ericm** [comment]
No problem

**Comment by ericm** [comment]
Should be resolved now :)

**Pull request merged**